documentation/puppet-api/v4/catalog.markdown

```diff
@@ -0,0 +1,76 @@
+---
+layout: default
+title: "Puppet Server: Puppet API: Catalog"
+canonical: "/puppetserver/latest/puppet-api/v4/catalog.html"
+---
+
+The catalog API returns a compiled catalog for the node specified in the request,
+making use of provided metadata like facts or environment if specified.
+If not specified, it will attempt to fetch this data from Puppet's configured sources
+(usually PuppetDB or a node classifier). The returned catalog is in JSON format,
+ready to be parsed and applied by an agent.
+
+## `POST /puppet/v4/catalog`
+
+(Introduced in Puppet Server 6.3.0)
+
+The input data for the catalog to be compiled is submitted as a JSON body with the 
+following form:
+```
+{
+  "certname": "<node name>",
+  "persistence": { "facts": <true/false>, "catalog": <true/false> },
+  # The rest are optional:
+  "facts": { "values": { "<fact name>": <fact value>, ... } },
+  "trusted_facts": { "values": { "<fact name>": <fact value>, ... } },
+  "environment": "<environment name>",
+  "options": { "prefer_requested_environment": <true/false>,
+               "capture_logs": <true/false> }
+}
+```
+
+#### `certname` (required)
+The name of the node for which to compile the catalog.
+
+#### `persistence` (required)
+A hash containing two required keys, `facts` and `catalog`, which when set to true will
+cause the facts and reports to be stored in PuppetDB, or discarded if set to false.
+
+#### `facts`
+A hash with a required `values` key, containing a hash of all the facts for the node.
+If not provided, Puppet will attempt to fetch facts for the node from PuppetDB.
+
+#### `trusted_facts`
+A hash with a required `values` key containing a hash of the trusted facts for a node.
```

Comment:
`facts` has `values` because that's how a facts object is encoded and you could include other data at the same level as `values`, but trusted facts aren't currently shaped like this. We could still keep it that way, just wanted to check

Reply:
I added the `values` key to `trusted_facts` also because we had talked about keeping our options open for adding adjacent keys to all hashes in our schema. That being said, there was already one bug we found snuck in around that difference in structure. So I could go either way.